File: src/crewai_multiagent_demo/gui/bridge.py
from __future__ import annotations

import json
import os
from collections.abc import Callable
from pathlib import Path
from typing import Any

from PySide6.QtCore import QObject, QThread, QUrl, Signal, Slot
from PySide6.QtGui import QDesktopServices

from crewai_multiagent_demo.config.loader import ConfigLoader
from crewai_multiagent_demo.core.runner import run_workflow
from crewai_multiagent_demo.domain.run_request import RunRequest
from crewai_multiagent_demo.gui.run_process import WorkflowProcessWorker
from crewai_multiagent_demo.gui.services import ConfigEditorService, HistoryService
from crewai_multiagent_demo.gui.state import RunStatus, event_label, progress_from_events
from crewai_multiagent_demo.llm.model_registry import MODEL_REGISTRY, ModelRegistry
from crewai_multiagent_demo.utils.diagnostics import create_diagnostics_archive
from crewai_multiagent_demo.utils.environment import has_api_key_for_model
from crewai_multiagent_demo.utils.files import atomic_write_json, atomic_write_text
from crewai_multiagent_demo.utils.paths import (
    BUNDLED_CONFIG_DIR,
    DEFAULT_CACHE_DIR,
    DEFAULT_CONFIG_DIR,
    DEFAULT_ENV_FILE,
    DEFAULT_OUTPUT_DIR,
)

PROTOCOL_VERSION = 1
WorkflowRunner = Callable[..., object]
ApiKeyChecker = Callable[[str], bool]


class InlineWorkflowWorker(QThread):
    """Test/development adapter; production runs use WorkflowProcessWorker."""

    eventReceived = Signal(dict)
    completed = Signal(object)
    failed = Signal(str)
    cancelled = Signal(str)

    def __init__(self, runner: WorkflowRunner, request: RunRequest) -> None:
        super().__init__()
        self._runner = runner
        self.request = request

    def request_cancel(self) -> None:
        # Injected test runners are deliberately in-process and cannot be
        # forcefully interrupted. The production path never uses this adapter.
        return

    def run(self) -> None:
        try:
            result = self._runner(
                topic=self.request.topic,
                model_alias=self.request.model_alias,
                config_dir=self.request.config_dir,
                output_dir=self.request.output_dir,
                run_id=self.request.run_id,
                request_timeout_seconds=self.request.request_timeout_seconds,
                max_retries=self.request.max_retries,
                on_event=self.eventReceived.emit,
            )
        except Exception as exc:  # pragma: no cover - surfaced through the bridge.
            self.failed.emit(str(exc))
            return
        self.completed.emit(result)


class StudioBridge(QObject):
    """Path-safe QWebChannel protocol adapter for the desktop UI."""

    runStateChanged = Signal(str)
    eventReceived = Signal(str)
    historyChanged = Signal(str)
    noticeRaised = Signal(str)

    def __init__(
        self,
        *,
        config_dir: Path = DEFAULT_CONFIG_DIR,
        output_dir: Path = DEFAULT_OUTPUT_DIR,
        env_file: Path = DEFAULT_ENV_FILE,
        cache_dir: Path = DEFAULT_CACHE_DIR,
        legacy_output_dir: Path | None = None,
        runner: WorkflowRunner = run_workflow,
        registry: ModelRegistry = MODEL_REGISTRY,
        api_key_checker: ApiKeyChecker = has_api_key_for_model,
    ) -> None:
        super().__init__()
        self.config_service = ConfigEditorService(config_dir)
        self.output_dir = Path(output_dir)
        self.env_file = Path(env_file)
        self.cache_dir = Path(cache_dir)
        self.legacy_output_dir = Path(legacy_output_dir) if legacy_output_dir else None
        self._runner = runner
        self._registry = registry
        self._api_key_checker = api_key_checker
        self._worker: WorkflowProcessWorker | InlineWorkflowWorker | None = None
        self._events: list[dict[str, Any]] = []
        self._last_result: object | None = None
        self._history_paths: dict[str, Path] = {}
        self._state: dict[str, Any] = self._empty_state()

    def is_running(self) -> bool:
        return self._state["status"] == RunStatus.RUNNING.value

    @Slot(result=str)
    def bootstrap(self) -> str:
        try:
            return self._json(self._response(data=self._snapshot()))
        except Exception as exc:
            return self._json(
                self._response(
                    ok=False,
                    code="bootstrap_failed",
                    message=str(exc),
                    data={
                        "protocolVersion": PROTOCOL_VERSION,
                        "configPath": str(self.config_service.config_dir),
                        "recoverable": True,
                    },
                )
            )

    @Slot(str, str, result=str)
    def startRun(self, topic: str, model_alias: str) -> str:  # noqa: N802
        if self.is_running():
            return self._json(self._response(ok=False, code="run_active", message="工作流正在运行。"))
        try:
            model = self._registry.resolve(model_alias)
            self.config_service.validate_current_config()
        except Exception as exc:
            return self._json(self._response(ok=False, code="invalid_config", message=str(exc)))
        if not self._api_key_checker(model.crewai_model):
            message = f"未检测到当前模型所需的 API key。请在设置中配置。配置文件：{self.env_file}"
            self._state = {**self._empty_state(), "status": RunStatus.FAILED.value, "error": message}
            self._emit_state()
            return self._json(self._response(ok=False, code="missing_api_key", message=message))

        request = RunRequest.create(
            topic=topic,
            model_alias=model.alias,
            config_dir=self.config_service.config_dir,
            output_dir=self.output_dir,
            env_file=self.env_file,
        )
        self._events = []
        self._last_result = None
        self._state = {
            **self._empty_state(),
            "status": RunStatus.RUNNING.value,
            "runId": request.run_id,
            "modelAlias": model.alias,
            "topic": topic.strip(),
            "taskCount": self.config_service.enabled_task_count(),
            "progress": 5,
            "activeAgent": "等待启动",
        }
        self._emit_state()

        if self._runner is run_workflow:
            worker: WorkflowProcessWorker | InlineWorkflowWorker = WorkflowProcessWorker(request)
        else:
            worker = InlineWorkflowWorker(self._runner, request)
        self._worker = worker
        worker.eventReceived.connect(self._handle_event)
        worker.completed.connect(self._handle_completed)
        worker.failed.connect(self._handle_failed)
        worker.cancelled.connect(self._handle_cancelled)
        worker.finished.connect(self._release_worker)
        worker.start()
        return self._json(self._response(data={"accepted": True, "runId": request.run_id}))

    @Slot(result=str)
    def cancelRun(self) -> str:  # noqa: N802
        if not self.is_running() or self._worker is None:
            return self._json(self._response(ok=False, code="run_inactive", message="当前没有正在运行的工作流。"))
        self._worker.request_cancel()
        self._state["activeAgent"] = "正在取消"
        self._emit_state()
        return self._json(self._response(data={"accepted": True}))

    def shutdown(self, wait_ms: int = 5500) -> None:
        if self._worker is None:
            return
        self._worker.request_cancel()
        self._worker.wait(wait_ms)

    @Slot(result=str)
    def resetRun(self) -> str:  # noqa: N802
        if self.is_running():
            return self._json(self._response(ok=False, code="run_active", message="运行结束前不能重置。"))
        self._events = []
        self._last_result = None
        self._state = self._empty_state()
        self._emit_state()
        return self._json(self._response(data=self._state))

    @Slot(result=str)
    def validateConfig(self) -> str:  # noqa: N802
        try:
            return self._json(self._response(data={"text": self.config_service.validation_text()}))
        except Exception as exc:
            return self._json(self._response(ok=False, code="invalid_config", message=str(exc)))

    @Slot(str, str, result=str)
    def saveConfig(self, kind: str, payload_json: str) -> str:  # noqa: N802
        if self.is_running():
            return self._json(self._response(ok=False, code="run_active", message="运行中不能修改配置。"))
        try:
            payload = json.loads(payload_json)
            if kind == "agents":
                self.config_service.save_agents_payload(payload)
            elif kind == "tasks":
                self.config_service.save_tasks_payload(payload)
            else:
                raise ValueError("未知配置类型")
        except Exception as exc:
            return self._json(self._response(ok=False, code="save_failed", message=str(exc)))
        return self._json(self._response(data=self._config_snapshot()))

    @Slot(str, str, result=str)
    def saveConfigBundle(self, agents_json: str, tasks_json: str) -> str:  # noqa: N802
        if self.is_running():
            return self._json(self._response(ok=False, code="run_active", message="运行中不能修改配置。"))
        try:
            self.config_service.save_config_payloads(json.loads(agents_json), json.loads(tasks_json))
        except Exception as exc:
            return self._json(self._response(ok=False, code="save_failed", message=str(exc)))
        return self._json(self._response(data=self._config_snapshot()))

    @Slot(result=str)
    def resetConfig(self) -> str:  # noqa: N802
        if self.is_running():
            return self._json(self._response(ok=False, code="run_active", message="运行中不能重置配置。"))
        try:
            document = ConfigLoader.load_json_file(BUNDLED_CONFIG_DIR / "workflow.json")
            atomic_write_json(self.config_service.workflow_file, document)
            return self._json(self._response(data=self._snapshot()))
        except Exception as exc:
            return self._json(self._response(ok=False, code="reset_failed", message=str(exc)))

    @Slot(str, result=str)
    def saveApiKey(self, key: str) -> str:  # noqa: N802
        cleaned = key.strip()
        if len(cleaned) < 8 or "\n" in cleaned or "\r" in cleaned:
            return self._json(self._response(ok=False, code="invalid_api_key", message="API key 格式无效。"))
        try:
            lines = self.env_file.read_text(encoding="utf-8").splitlines() if self.env_file.exists() else []
            updated: list[str] = []
            replaced = False
            for line in lines:
                if line.startswith("DEEPSEEK_API_KEY="):
                    updated.append(f"DEEPSEEK_API_KEY={cleaned}")
                    replaced = True
                else:
                    updated.append(line)
            if not replaced:
                updated.append(f"DEEPSEEK_API_KEY={cleaned}")
            atomic_write_text(self.env_file, "\n".join(updated) + "\n")
            os.environ["DEEPSEEK_API_KEY"] = cleaned
            return self._json(self._response(data={"configured": True, "envFile": str(self.env_file)}))
        except Exception as exc:
            return self._json(self._response(ok=False, code="save_key_failed", message=str(exc)))

    @Slot(result=str)
    def exportDiagnostics(self) -> str:  # noqa: N802
        try:
            archive = create_diagnostics_archive(self.cache_dir, self.output_dir)
            QDesktopServices.openUrl(QUrl.fromLocalFile(str(archive.parent)))
            return self._json(self._response(data={"path": str(archive)}))
        except Exception as exc:
            return self._json(self._response(ok=False, code="diagnostics_failed", message=str(exc)))

    @Slot(str, result=str)
    def loadHistory(self, record_id: str) -> str:  # noqa: N802
        self._refresh_history_paths()
        path = self._history_paths.get(record_id)
        if path is None:
            return self._json(self._response(ok=False, code="history_missing", message="找不到该历史记录。"))
        selection = HistoryService(path.parent).load_selection(path)
        return self._json(
            self._response(
                data={
                    "id": record_id,
                    "summary": selection.summary,
                    "fullReport": selection.full_report,
                    "metadata": selection.metadata,
                }
            )
        )

    @Slot(str, str, result=str)
    def openLocation(self, target: str, record_id: str = "") -> str:  # noqa: N802
        path: Path | None
        if target == "outputs":
            path = self.output_dir
        elif target == "config":
            path = self.config_service.config_dir
        elif target == "history":
            self._refresh_history_paths()
            path = self._history_paths.get(record_id)
        else:
            path = None
        if path is None or not path.exists():
            return self._json(self._response(ok=False, code="location_missing", message="没有可打开的位置。"))
        opened = QDesktopServices.openUrl(QUrl.fromLocalFile(str(path)))
        return self._json(self._response(ok=opened, code="open_failed" if not opened else None))

    def _handle_event(self, event: dict[str, Any]) -> None:
        public_event = dict(event)
        output = str(public_event.pop("output", ""))
        if output:
            public_event["outputPreview"] = output[:240]
        self._events.append(public_event)
        event_type = str(public_event.get("type", ""))
        self._state["progress"] = progress_from_events(self._events, self._state["taskCount"])
        self._state["activeAgent"] = str(
            public_event.get("task_name") or public_event.get("agent") or event_label(event_type)
        )
        self._state["events"] = self._events[-50:]
        self.eventReceived.emit(self._json(public_event))
        self._emit_state()

    def _handle_completed(self, result: object) -> None:
        self._last_result = result
        run_dir = Path(str(getattr(result, "run_dir", "")))
        self._state = {
            **self._state,
            "status": RunStatus.SUCCEEDED.value,
            "progress": 100,
            "result": {
                "historyId": self._id_for_path(run_dir),
                "modelAlias": str(getattr(result, "model_alias", self._state["modelAlias"])),
                "elapsedSeconds": float(getattr(result, "elapsed_seconds", 0.0)),
                "summary": str(getattr(result, "concise_report", "")),
                "fullReport": str(getattr(result, "full_report", "")),
                "taskOutputs": list(getattr(result, "task_outputs", []) or []),
            },
        }
        self._emit_state()
        self.historyChanged.emit(self._json(self._history_items()))
        self.noticeRaised.emit(self._json({"kind": "success", "message": "工作流已完成，报告已生成。"}))

    def _handle_failed(self, error: str) -> None:
        self._state = {**self._state, "status": RunStatus.FAILED.value, "progress": 0, "error": error}
        self._emit_state()
        self.historyChanged.emit(self._json(self._history_items()))
        self.noticeRaised.emit(self._json({"kind": "error", "message": error}))

    def _handle_cancelled(self, message: str) -> None:
        self._state = {
            **self._state,
            "status": RunStatus.CANCELLED.value,
            "progress": 0,
            "activeAgent": "已取消",
            "error": message,
        }
        self._emit_state()
        self.historyChanged.emit(self._json(self._history_items()))
        self.noticeRaised.emit(self._json({"kind": "warning", "message": message}))

    def _release_worker(self) -> None:
        if self._worker is not None:
            self._worker.deleteLater()
        self._worker = None

    def _snapshot(self) -> dict[str, Any]:
        return {
            "protocolVersion": PROTOCOL_VERSION,
            "models": list(self._registry.aliases),
            "defaultModel": self._registry.default_alias(),
            "apiKeyConfigured": {
                alias: self._api_key_checker(self._registry.resolve(alias).crewai_model)
                for alias in self._registry.aliases
            },
            "configPath": str(self.config_service.config_dir),
            "envFile": str(self.env_file),
            "config": self._config_snapshot(),
            "history": self._history_items(),
            "runState": self._state,
        }

    def _config_snapshot(self) -> dict[str, Any]:
        snapshot = self.config_service.load_snapshot()
        return {
            "agents": snapshot.agents,
            "tasks": snapshot.tasks,
            "agentsJson": snapshot.agents_json,
            "tasksJson": snapshot.tasks_json,
            "enabledTaskCount": snapshot.enabled_task_count,
            "validationText": snapshot.validation_text,
        }

    def _history_items(self) -> list[dict[str, Any]]:
        self._refresh_history_paths()
        return [
            HistoryService(path.parent).item_for_path(record_id, path)
            for record_id, path in self._history_paths.items()
        ]

    def _refresh_history_paths(self) -> None:
        paths: dict[str, Path] = {}
        for prefix, root in (("current", self.output_dir), ("legacy", self.legacy_output_dir)):
            if root is None:
                continue
            for path in HistoryService(root).list_run_dirs():
                paths[f"{prefix}:{path.name}"] = path
        self._history_paths = dict(sorted(paths.items(), key=lambda item: item[1].name, reverse=True))

    def _id_for_path(self, path: Path) -> str:
        self._refresh_history_paths()
        for record_id, candidate in self._history_paths.items():
            if candidate == path:
                return record_id
        return f"current:{path.name}"

    @staticmethod
    def _empty_state() -> dict[str, Any]:
        return {
            "status": RunStatus.IDLE.value,
            "runId": "",
            "progress": 0,
            "modelAlias": MODEL_REGISTRY.default_alias(),
            "topic": "",
            "taskCount": 0,
            "activeAgent": "等待启动",
            "events": [],
            "result": None,
            "error": None,
        }

    def _emit_state(self) -> None:
        self.runStateChanged.emit(self._json(self._state))

    @staticmethod
    def _response(
        *, ok: bool = True, data: Any | None = None, code: str | None = None, message: str | None = None
    ) -> dict[str, Any]:
        response: dict[str, Any] = {"ok": ok}
        if data is not None:
            response["data"] = data
        if code:
            response["code"] = code
        if message:
            response["message"] = message
        return response

    @staticmethod
    def _json(value: Any) -> str:
        return json.dumps(value, ensure_ascii=False, default=str)

File: frontend/src/bridge.ts
import type {
  AgentConfig,
  ApiResponse,
  AppSnapshot,
  ConfigSnapshot,
  HistoryDetail,
  HistoryItem,
  Notice,
  RunState,
  TaskConfig
} from "./types";

const PROTOCOL_VERSION = 2;
type Listener<T> = (value: T) => void;
type Unsubscribe = () => void;

export interface StudioBridge {
  bootstrap(): Promise<ApiResponse<AppSnapshot>>;
  startRun(topic: string, model: string): Promise<ApiResponse<{ accepted: boolean; runId: string }>>;
  cancelRun(): Promise<ApiResponse<{ accepted: boolean }>>;
  resetRun(): Promise<ApiResponse<RunState>>;
  validateConfig(): Promise<ApiResponse<{ text: string }>>;
  saveConfig(kind: "agents" | "tasks", payload: AgentConfig[] | TaskConfig[], baseRevision?: string): Promise<ApiResponse<ConfigSnapshot>>;
  saveConfigBundle(agents: AgentConfig[], tasks: TaskConfig[], baseRevision?: string): Promise<ApiResponse<ConfigSnapshot>>;
  saveGraph(graph: ConfigSnapshot["graph"], baseRevision?: string): Promise<ApiResponse<ConfigSnapshot>>;
  resetConfig(): Promise<ApiResponse<AppSnapshot>>;
  saveApiKey(key: string): Promise<ApiResponse<{ configured: boolean; envFile: string }>>;
  exportDiagnostics(): Promise<ApiResponse<{ path: string }>>;
  loadHistory(id: string): Promise<ApiResponse<HistoryDetail>>;
  openLocation(target: "outputs" | "history" | "config", id?: string): Promise<ApiResponse<unknown>>;
  onRunState(listener: Listener<RunState>): Unsubscribe;
  onHistory(listener: Listener<HistoryItem[]>): Unsubscribe;
  onNotice(listener: Listener<Notice>): Unsubscribe;
}

type ChannelSignal = { connect(callback: (json: string) => void): void; disconnect(callback: (json: string) => void): void };
type ChannelObject = Record<string, (...args: unknown[]) => void> & {
  runStateChanged: ChannelSignal;
  historyChanged: ChannelSignal;
  noticeRaised: ChannelSignal;
};

declare global {
  interface Window {
    qt?: { webChannelTransport: unknown };
    QWebChannel?: new (transport: unknown, ready: (channel: { objects: { studio: ChannelObject } }) => void) => unknown;
  }
}

function decode<T>(value: string): ApiResponse<T> {
  try {
    return JSON.parse(value) as ApiResponse<T>;
  } catch {
    return { ok: false, code: "bridge_decode", message: "桌面通信返回了无效数据。" };
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value && typeof value === "object" && !Array.isArray(value));
}

function isGraphLayout(value: unknown): value is ConfigSnapshot["graph"] {
  if (!isRecord(value) || !isRecord(value.positions) || !isRecord(value.viewport)) return false;
  const viewport = value.viewport;
  return typeof viewport.x === "number" && Number.isFinite(viewport.x)
    && typeof viewport.y === "number" && Number.isFinite(viewport.y)
    && typeof viewport.zoom === "number" && Number.isFinite(viewport.zoom)
    && Object.values(value.positions).every((position) => isRecord(position)
      && typeof position.x === "number" && Number.isFinite(position.x)
      && typeof position.y === "number" && Number.isFinite(position.y));
}

function isConfigSnapshot(value: unknown): value is ConfigSnapshot {
  if (!isRecord(value)) return false;
  return Array.isArray(value.agents)
    && Array.isArray(value.tasks)
    && isGraphLayout(value.graph)
    && typeof value.revision === "string"
    && typeof value.agentsJson === "string"
    && typeof value.tasksJson === "string"
    && typeof value.enabledTaskCount === "number"
    && typeof value.validationText === "string";
}

function isRunState(value: unknown): value is RunState {
  if (!isRecord(value) || !Array.isArray(value.events) || !isRecord(value.taskStates)) return false;
  return typeof value.status === "string" && typeof value.runId === "string"
    && typeof value.progress === "number" && typeof value.modelAlias === "string"
    && typeof value.topic === "string" && typeof value.taskCount === "number"
    && typeof value.activeAgent === "string";
}

function isNotice(value: unknown): value is Notice {
  return isRecord(value) && typeof value.kind === "string" && typeof value.message === "string";
}

function isAppSnapshot(value: unknown): value is AppSnapshot {
  if (!isRecord(value)) return false;
  const row = value;
  const config = row.config;
  return row.protocolVersion === PROTOCOL_VERSION
    && Array.isArray(row.models)
    && typeof row.defaultModel === "string"
    && isConfigSnapshot(config)
    && Array.isArray(row.history)
    && isRunState(row.runState);
}

function subscribe<T>(signal: ChannelSignal, listener: Listener<T>, guard?: (value: unknown) => value is T): Unsubscribe {
  const callback = (json: string) => {
    try {
      const value: unknown = JSON.parse(json);
      if (!guard || guard(value)) listener(value as T);
    } catch { /* A later full state can recover. */ }
  };
  signal.connect(callback);
  return () => signal.disconnect(callback);
}

function configResponse(response: ApiResponse<ConfigSnapshot>): ApiResponse<ConfigSnapshot> {
  if (response.ok && !isConfigSnapshot(response.data)) {
    return { ok: false, code: "protocol_mismatch", message: "桌面配置响应不符合当前协议。" };
  }
  return response;
}

class QtChannelBridge implements StudioBridge {
  constructor(private readonly studio: ChannelObject) {}

  private invoke<T>(method: string, ...args: unknown[]): Promise<ApiResponse<T>> {
    return new Promise((resolve) => {
      const target = this.studio?.[method];
      if (typeof target !== "function") {
        resolve({ ok: false, code: "bridge_method_missing", message: `桌面后端缺少方法：${method}` });
        return;
      }
      target(...args, (result: unknown) => resolve(decode<T>(String(result))));
    });
  }

  async bootstrap() {
    const response = await this.invoke<AppSnapshot>("bootstrap");
    if (response.ok && !isAppSnapshot(response.data)) {
      return { ok: false, code: "protocol_mismatch", message: "桌面前后端协议版本不兼容。" };
    }
    return response;
  }
  startRun(topic: string, model: string) { return this.invoke<{ accepted: boolean; runId: string }>("startRun", topic, model); }
  cancelRun() { return this.invoke<{ accepted: boolean }>("cancelRun"); }
  resetRun() { return this.invoke<RunState>("resetRun"); }
  validateConfig() { return this.invoke<{ text: string }>("validateConfig"); }
  saveConfig(kind: "agents" | "tasks", payload: AgentConfig[] | TaskConfig[], baseRevision = "") {
    return this.invoke<ConfigSnapshot>("saveConfig", kind, JSON.stringify(payload), baseRevision).then(configResponse);
  }
  saveConfigBundle(agents: AgentConfig[], tasks: TaskConfig[], baseRevision = "") {
    return this.invoke<ConfigSnapshot>("saveConfigBundle", JSON.stringify(agents), JSON.stringify(tasks), baseRevision).then(configResponse);
  }
  saveGraph(graph: ConfigSnapshot["graph"], baseRevision = "") {
    return this.invoke<ConfigSnapshot>("saveGraph", JSON.stringify(graph), baseRevision).then(configResponse);
  }
  resetConfig() { return this.invoke<AppSnapshot>("resetConfig"); }
  saveApiKey(key: string) { return this.invoke<{ configured: boolean; envFile: string }>("saveApiKey", key); }
  exportDiagnostics() { return this.invoke<{ path: string }>("exportDiagnostics"); }
  loadHistory(id: string) { return this.invoke<HistoryDetail>("loadHistory", id); }
  openLocation(target: "outputs" | "history" | "config", id = "") { return this.invoke("openLocation", target, id); }
  onRunState(listener: Listener<RunState>) { return subscribe(this.studio.runStateChanged, listener, isRunState); }
  onHistory(listener: Listener<HistoryItem[]>) { return subscribe(this.studio.historyChanged, listener, (value): value is HistoryItem[] => Array.isArray(value)); }
  onNotice(listener: Listener<Notice>) { return subscribe(this.studio.noticeRaised, listener, isNotice); }
}

const demoAgents: AgentConfig[] = [
  { id: "problem_analyst", role: "问题分析师", goal: "梳理问题", backstory: "结构化分析", enabled: true },
  { id: "solution_strategist", role: "方案策略师", goal: "形成方案", backstory: "务实设计", enabled: true }
];
const demoTasks: TaskConfig[] = [
  { id: "analysis", name: "问题分析", description: "分析 {topic}", expected_output: "结构化分析", agent_id: "problem_analyst", context_task_ids: [], artifact_role: "full_report", enabled: true },
  { id: "summary", name: "精简总结", description: "总结结果", expected_output: "精简报告", agent_id: "solution_strategist", context_task_ids: ["analysis"], artifact_role: "summary", enabled: true }
];
const demoGraph: ConfigSnapshot["graph"] = {
  positions: { analysis: { x: 40, y: 90 }, summary: { x: 310, y: 90 } },
  viewport: { x: 0, y: 0, zoom: 1 }
};

function initialState(): RunState {
  return { status: "idle", runId: "", progress: 0, modelAlias: "flash", topic: "", taskCount: 2, activeAgent: "等待启动", events: [], taskStates: {}, result: null, error: null };
}

class MockBridge implements StudioBridge {
  private state = initialState();
  private history: HistoryItem[] = [];
  private agents = demoAgents;
  private tasks = demoTasks;
  private graph = demoGraph;
  private stateListeners = new Set<Listener<RunState>>();
  private historyListeners = new Set<Listener<HistoryItem[]>>();
  private noticeListeners = new Set<Listener<Notice>>();
  private timers: number[] = [];
  private config(): ConfigSnapshot { return { agents: this.agents, tasks: this.tasks, graph: this.graph, revision: "demo", agentsJson: JSON.stringify(this.agents, null, 2), tasksJson: JSON.stringify(this.tasks, null, 2), enabledTaskCount: this.tasks.filter((task) => task.enabled).length, validationText: "开发模式模拟配置。" }; }
  private snapshot(): AppSnapshot { return { protocolVersion: PROTOCOL_VERSION, models: ["flash", "pro"], defaultModel: "flash", apiKeyConfigured: { flash: true, pro: true }, configPath: "development/mock", envFile: "development/mock/.env", config: this.config(), history: this.history, runState: this.state }; }
  private emitState() { this.stateListeners.forEach((listener) => listener({ ...this.state })); }
  private emitHistory() { this.historyListeners.forEach((listener) => listener([...this.history])); }
  private notice(notice: Notice) { this.noticeListeners.forEach((listener) => listener(notice)); }
  async bootstrap(): Promise<ApiResponse<AppSnapshot>> { return { ok: true, data: this.snapshot() }; }
  async startRun(topic: string, model: string): Promise<ApiResponse<{ accepted: boolean; runId: string }>> {
    if (this.state.status === "running") return { ok: false, code: "run_active", message: "工作流正在运行。" };
    const runId = `demo-${Date.now()}`;
    this.state = { ...initialState(), status: "running", runId, modelAlias: model, topic, taskCount: this.tasks.filter((task) => task.enabled).length, progress: 5, activeAgent: "问题分析师" };
    this.emitState();
    this.tasks.forEach((task, index) => this.timers.push(window.setTimeout(() => {
      const event = { type: "task_completed", task_id: task.id, task_name: task.name, agent: task.agent_id };
      this.state = { ...this.state, events: [...this.state.events, event], activeAgent: task.name, progress: Math.min(95, 10 + (index + 1) * 40) };
      this.emitState();
    }, 450 * (index + 1))));
    this.timers.push(window.setTimeout(() => {
      const record: HistoryItem = { id: runId, createdAt: new Date().toISOString(), modelAlias: model, elapsedSeconds: 1.4, status: "succeeded" };
      this.history = [record, ...this.history];
      this.state = { ...this.state, status: "succeeded", progress: 100, result: { historyId: record.id, modelAlias: model, elapsedSeconds: 1.4, summary: `已完成“${topic || "默认主题"}”。`, fullReport: `# 开发模式报告\n\n${topic}`, taskOutputs: [] } };
      this.emitState(); this.emitHistory(); this.notice({ kind: "success", message: "开发模式模拟运行已完成。" });
    }, 1500));
    return { ok: true, data: { accepted: true, runId } };
  }
  async cancelRun() { this.timers.forEach(window.clearTimeout); this.timers = []; this.state = { ...this.state, status: "cancelled", progress: 0 }; this.emitState(); return { ok: true, data: { accepted: true } }; }
  async resetRun() { this.state = initialState(); this.emitState(); return { ok: true, data: this.state }; }
  async validateConfig() { return { ok: true, data: { text: this.config().validationText } }; }
  async saveConfig(kind: "agents" | "tasks", payload: AgentConfig[] | TaskConfig[], _baseRevision = "") { if (kind === "agents") this.agents = payload as AgentConfig[]; else this.tasks = payload as TaskConfig[]; return { ok: true, data: this.config() }; }
  async saveConfigBundle(agents: AgentConfig[], tasks: TaskConfig[], _baseRevision = "") { this.agents = agents; this.tasks = tasks; return { ok: true, data: this.config() }; }
  async saveGraph(graph: ConfigSnapshot["graph"], _baseRevision = "") { this.graph = graph; return { ok: true, data: this.config() }; }
  async resetConfig() { this.agents = demoAgents; this.tasks = demoTasks; this.graph = demoGraph; return { ok: true, data: this.snapshot() }; }
  async saveApiKey() { return { ok: true, data: { configured: true, envFile: "development/mock/.env" } }; }
  async exportDiagnostics() { return { ok: true, data: { path: "development/mock/diagnostics.zip" } }; }
  async loadHistory(id: string) { return { ok: true, data: { id, summary: "模拟摘要", fullReport: "# 模拟报告", metadata: "开发模式" } }; }
  async openLocation() { return { ok: true }; }
  onRunState(listener: Listener<RunState>) { this.stateListeners.add(listener); return () => this.stateListeners.delete(listener); }
  onHistory(listener: Listener<HistoryItem[]>) { this.historyListeners.add(listener); return () => this.historyListeners.delete(listener); }
  onNotice(listener: Listener<Notice>) { this.noticeListeners.add(listener); return () => this.noticeListeners.delete(listener); }
}

class UnavailableBridge implements StudioBridge {
  private error<T>(): Promise<ApiResponse<T>> { return Promise.resolve({ ok: false, code: "desktop_bridge_unavailable", message: "桌面后端未连接。请重新安装或检查应用日志。" }); }
  bootstrap() { return this.error<AppSnapshot>(); }
  startRun() { return this.error<{ accepted: boolean; runId: string }>(); }
  cancelRun() { return this.error<{ accepted: boolean }>(); }
  resetRun() { return this.error<RunState>(); }
  validateConfig() { return this.error<{ text: string }>(); }
  saveConfig() { return this.error<ConfigSnapshot>(); }
  saveConfigBundle() { return this.error<ConfigSnapshot>(); }
  saveGraph() { return this.error<ConfigSnapshot>(); }
  resetConfig() { return this.error<AppSnapshot>(); }
  saveApiKey() { return this.error<{ configured: boolean; envFile: string }>(); }
  exportDiagnostics() { return this.error<{ path: string }>(); }
  loadHistory() { return this.error<HistoryDetail>(); }
  openLocation() { return this.error<unknown>(); }
  onRunState() { return () => {}; }
  onHistory() { return () => {}; }
  onNotice() { return () => {}; }
}

let bridgePromise: Promise<StudioBridge> | undefined;
export function getBridge(): Promise<StudioBridge> {
  if (bridgePromise) return bridgePromise;
  bridgePromise = new Promise((resolve) => {
    if (window.qt?.webChannelTransport && window.QWebChannel) {
      new window.QWebChannel(window.qt.webChannelTransport, (channel) => resolve(new QtChannelBridge(channel.objects.studio)));
      return;
    }
    resolve(import.meta.env.DEV ? new MockBridge() : new UnavailableBridge());
  });
  return bridgePromise;
}
